### PR TITLE
Refactor: Create Enum for custom Amp Events

### LINKED ISSUE #9738: Enum for AMP events
### First Timers Only
We know figuring out the process for contributing to an open source project can be intimidating, so we created this issue as a way for you to learn the ropes.  (If you feel comfortable contributing to open source projects, please leave this issue for someone else.)

### What you will need to know

- Some familiarity with JavaScript.

### Background

The DOM has a [rich event model](https://developer.mozilla.org/en-US/docs/Web/API/Event), e.g. for handling user input. AMP also uses custom events for coordinating behavior across components and modules in the code base. These AMP events follow a naming convention being prefixed with `amp:`, e.g. [`amp:built`](https://github.com/ampproject/amphtml/search?utf8=%E2%9C%93&q=%22amp%3Abuilt%22&type=).

### Motivation

Most usages of `amp:` events today use the string literal directly, for example [on this line](https://github.com/ampproject/amphtml/blob/2699e24df5a1ea47dcc143779c7eadf4e95a3731/src/service/resource.js#L311).

```
this.element.dispatchCustomEvent('amp:built');
```

This is nice and simple, but it results in duplication of the `amp:built` string and has the risk of typos. A better way to do this is to define these events in an enum, like what [video-interface#VideoEvents](https://github.com/ampproject/amphtml/blob/2699e24df5a1ea47dcc143779c7eadf4e95a3731/src/video-interface.js#L154) does.

### The bug

Let's create a new file, `src/amp-events.js`, that contains an enum defining the commonly used events for AMP components. This will reduce duplication, eliminate the risk of typos, and improve readability of our code base.

### Step by step
- [x] Claim this issue by adding a comment below.  Please only claim this bug if you plan on starting work in the next day or so.
- [ ] If you aren't too familiar with Git/GitHub, see the [Getting Started End-to-End Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md) for [an intro to Git & GitHub,](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#intro-to-git-and-github) and [how to get a copy of the code](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#get-a-copy-of-the-amphtml-code).  You can also refer to the [Quick Start Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-quick.md) for the necessary setup steps with less explanation than the End-to-End guide.
- [ ] Follow the instructions for [building AMP](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#building-amp-and-starting-a-local-server).
- [ ] [Create a Git branch](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#create-a-git-branch) for making your changes.
- [ ] Search the code base for AMP events (try searching for `'amp:`).
- [ ] Create the `src/amp-events.js` file with an object literal enumerating the events you've found in your search.
- [ ] Import your new file into files containing the existing string literal event names and replace them with a reference to your enum.
- [ ] [Commit your changes](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#edit-files-and-commit-them) frequently.
- [ ] [Push your changes to GitHub](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#push-your-changes-to-your-github-fork).
- [ ] [Create a Pull Request](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#send-a-pull-request-ie-request-a-code-review).  Mention `Fixes #9738` in the description and mention me (@choumx) to get a code review.
- [ ] [Respond to your reviewer's comments](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#respond-to-pull-request-comments) (if any).

Once approved, your changes will be merged.  **⚡⚡⚡Congrats on making your first contribution to the AMP Project!⚡⚡⚡**  You'll be able to see it [live across the web soon](https://github.com/ampproject/amphtml/blob/master/contributing/release-schedule.md)!

Thanks, and we hope to see more contributions from you soon.

### Questions?

If you have questions ask in this issue or on your Pull Request (if you've created one) or see the [How to get help](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#how-to-get-help) section of the Getting Started guide. Mention me (@choumx) so I can help!

### PR DESCRIPTION
**Pull request by abc516**
Fixes #9738 
@choumx

### CODE REVIEW

**Comment by abc516** [comment]
I signed the Google CLA!

File: src/amp-events.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+*
+* @constant {!Object<string, string>}
+*/
+export const AmpEvents = {
+  VISIBILITY_CHANGE: 'amp:visibilitychange',
+  BIND: {
+    INITIALIZE: 'amp:bind:initialize',
+    RESCAN_TEMPLATE: 'amp:bind:rescan-template',
+    SET_STATE: 'amp:bind:setState',
+  },
```
Comment (choumx) [inline src/amp-events.js]: These actually come from the AMP extension `amp-bind`, whose events should probably live in its own file under the `extensions/amp-bind/0.1/` directory. Please either exclude Bind events from this PR or create a new `BindEvents` enum and file in that directory. 

Sorry I didn't mention this in the issue. Nice job being thorough!

File: src/amp-events.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  TEMPLATE_RENDERED: 'amp:template-rendered',
+  DOM_UPDATE: 'amp:dom-update',
+  BUILT: 'amp:built',
+  ATTACHED: 'amp:attached',
+  STUBBED: 'amp:stubbed',
+  LOAD: {
+    START: 'amp:load:start',
+    END: 'amp:load:end',
```
Comment (choumx) [inline src/amp-events.js]: I think the type checker will complain about this. Let's keep the enum single-level: `LOAD_START` and `LOAD_END`.

File: src/amp-events.js (hunk 1 above)
Comment (abc516) [inline src/amp-events.js]: @choumx I'll create a BindEvents enum for this.

File: src/amp-events.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/**
+* Enum used to specify custom Amp Events
+*
+* @constant {!Object<string, string>}
```
Comment (abc516) [inline src/amp-events.js]: Cool, looks like an easy fix!

File: src/amp-events.js (hunk 2 above)
Comment (abc516) [inline src/amp-events.js]: Ok, I will make them single level.

**Comment by abc516** [comment]
@choumx I rebased my branch and now the googlebot removed the CLA approved status. What should I do?

**Comment by choumx** [comment]
@abc516 Looks like the rebase pulled in a few commits from other people. You can try to `git rebase -i master` again and remove those commits from the log. Recommend backing up your work first before using `rebase` to delete history.

**Comment by choumx** [comment]
Sorry about the merge conflict from #9788. 😅  Almost there!

**Comment by abc516** [comment]
@zhouyx @jridgewell @rsimha-amp Can you comment on this pull request and confirm you are ok with your commits being used in this Pull Request?

File: src/amp-events.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+=======
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+>>>>>>> fix: add Apache License comment to src/amp-events.js
+=======
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+>>>>>>> refactor: create separate Bind Events enum file
+=======
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+>>>>>>> 8bc197821b7802c0b7796cf9fee34dbf9c658706
```
Comment (choumx) [inline src/amp-events.js]: Please remove merge markers here.

File: test/integration/test-amp-ad-doubleclick.js [hunk 5]
```diff
@@ -122,7 +123,8 @@ describe.configure().retryOnSaucelabs().run('Rendering of one ad', () => {
     }).then(() => {
       expect(iframe.contentWindow.context.hidden).to.be.false;
       return new Promise(resolve => {
-        iframe.contentWindow.addEventListener('amp:visibilitychange', resolve);
+        iframe.contentWindow.addEventListener(
+          AmpEvents.VISIBILITY_CHANGE, resolve);
```
Comment (choumx) [inline test/integration/test-amp-ad-doubleclick.js]: Minor: Increase indent +2 spaces.

File: testing/iframe.js [hunk 6]
```diff
@@ -64,14 +66,14 @@ export function createFixtureIframe(fixture, initialIframeHeight, opt_beforeLoad
...
       'amp:form-service:initialize': 0,
-      'amp:load:start': 0,
-      'amp:stubbed': 0,
+      [AmpEvents.ATTACHED]: 0,
+      [BindEvents.INITIALIZE]: 0,
+      [BindEvents.SET_STATE]: 0,
+      [BindEvents.RESCAN_TEMPLATE]: 0,
+      [AmpEvents.ERROR]: 0,
+      [AmpEvents.LOAD_START]: 0,
+      [AmpEvents.STUBBED]: 0,
```
Comment (choumx) [inline testing/iframe.js]: Why wrap each event in an array?

Comment (abc516) [inline testing/iframe.js]: It's a computed property name, a feature of ES6. http://es6-features.org/#ComputedPropertyNames

File: test/integration/test-amp-ad-doubleclick.js (hunk 5 above)
Comment (abc516) [inline test/integration/test-amp-ad-doubleclick.js]: so +2 to lines 126 and 127?

File: src/amp-events.js (hunk 4 above)
Comment (abc516) [inline src/amp-events.js]: done

File: test/integration/test-amp-ad-doubleclick.js (hunk 5 above)
Comment (choumx) [inline test/integration/test-amp-ad-doubleclick.js]: Just 127.

Comment (rsimha-amp) [inline test/integration/test-amp-ad-doubleclick.js]: Just line 127. (Continuing on the next line requires 4 spaces of indent.)

**Comment by choumx** [comment]
Please whitelist `amp-events.js` in the relevant parts of `dep-check-config.js` to fix the presubmit error.

**Comment by abc516** [comment]
Thanks @choumx !

**Pull request merged**